raft-tech/TANF-app
# Issue 429: Add component to input single file

### Summary of Changes

- Added UploadReport component to Upload page
	- Added 4 required UploadFile sections
- Added redux actions and reducers and modeled File Upload states
- Added error handling for uploads of invalid file types
- Added placeholder 'Submit Files' button for uploads

_Addresses issue_ #429
_Acceptance criteria as stated in the issue_

### How to Test

1. Log in as any user.
2. Click the Reports tab.
3. Select a year in the Fiscal Year dropdown.
4. Click Search.
5. Ensure a new UploadReport section appears below the ~~Begin Report~~ Search button with the four required sections on the page.
6. Drag a text file (`.txt` file extension) to any of the four sections.
7. Check that there were no errors, and the input area changes to display a file icon with the name of the uploaded file.
8. Try the same with a non-`.txt` file, and check that an error message is shown 
9. Clicking Submit Files should just refresh the page.
10. Clicking Cancel or changing the selected year should remove the UploadReport form.

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] Files Input accepts a single file 
  - [x] Input accepts only -.txt file types
  - [x] Upload button is a placeholder
  - [x] Input displays a helpful error message when a wrong file type is uploaded 
  - [x] Frontend matches #427 delivered and accepted mockup


As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [x] Did automated and manual testing with @reitermb  using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

Accessibility issues have been addressed

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*


No security issues detected


<img src="https://user-images.githubusercontent.com/12901172/110814898-e3c47200-8257-11eb-984d-ba5878476edc.png " width="600" >
<img src="https://user-images.githubusercontent.com/12901172/110814899-e4f59f00-8257-11eb-8598-af60246b57b6.png" width="600">
<img src="https://user-images.githubusercontent.com/12901172/110814903-e4f59f00-8257-11eb-9d64-fa2339acd69b.png" width="600">
